### PR TITLE
Normalize report partitioning with new utility

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #3348

## Relevant technical choices

* Promoted to general utility as the same partitioning was done in a number of other places, not just in Search Console
* Updated all instances of previous array slicing to use the new utility

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by aaemnnosttv** [comment]
VRT failed with one widget rendering a slightly different date range in the chart although stats remained exactly the same so it looks like this may have been a bug before.

| Before | After |
| --- | --- |
| image | image |

image

File: assets/js/modules/search-console/components/module/ModuleOverviewWidget/index.js [hunk 1]
```diff
@@ -87,6 +89,8 @@ const ModuleOverviewWidget = ( { Widget, WidgetReportZero, WidgetReportError } )
 		);
 	}
 
+	const { compareRange, currentRange } = partitionReport( data, { rangeLength: dateRangeLength } );
```
Comment (felixarntz) [inline assets/js/modules/search-console/components/module/ModuleOverviewWidget/index.js:92]: Since this is only needed in the `Stats` component, why not do it there?

File: assets/js/util/partition-report.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	invariant( Number.isInteger( rangeLength ) && rangeLength > 0, 'rangeLength must be a positive integer.' );
+
+	const compareRange = report.slice( 0, rangeLength );
+	const currentRange = report.slice( rangeLength, rangeLength * 2 );
+	// Normalize the length of both ranges to be the same.
+	const minLength = Math.min( compareRange.length, currentRange.length );
+
+	return {
+		compareRange: compareRange.slice( 0, minLength ),
+		currentRange: currentRange.slice( 0, minLength ),
```
Comment (felixarntz) [inline assets/js/util/partition-report.js:45]: Could this result in a potential problem that the compare date range would be incorrect (e.g. missing some days)? For example, if the `currentRange` is 1 entry longer than `compareRange`, right now this would strip the last result of the `compareRange`, so effectively that date would be skipped which would result in somewhat of a wrong comparison (where furthermore even for a "multitude of a week" period weekdays may not be aligned).

A more appropriate solution may be to slice the `compareRange` from its end, effectively trimming away the oldest dates. This may be more specific to Site Kit's behavior, but I think that makes sense - we don't need this utility to be overly generic.

File: assets/js/util/partition-report.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * Splits report data into segments for current and comparison ranges.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Array}  report           Report rows.
+ * @param {Object} args             Additional arguments.
+ * @param {number} args.rangeLength Length of an individual range.
+ * @return {Object} Object with keys for `compareRange` and `currentRange`.
+ */
+export const partitionReport = ( report, { rangeLength } ) => {
```
Comment (felixarntz) [inline assets/js/util/partition-report.js:34]: While the code for this utility function right now appears to be quite generic, I'm not sure that's really necessary (see also other comment). It's somewhat of a nit-pick, but it may be better to call the argument here for example `dateRangeLength` for more clarity on what it refers to. The return object properties could possibly be something like `dateRangeEntries` and `compareDateRangeEntries`, just throwing some ideas out there.

TLDR I think it's a good idea to name and document this more focused around date ranges since that's what we actually use it for.

**Review by felixarntz — COMMENTED** [review]
This looks solid mostly, I have a few points of feedback, particularly one thing is worth improving.

File: assets/js/modules/search-console/components/module/ModuleOverviewWidget/index.js (hunk 1 above)
Comment (aaemnnosttv) [reply assets/js/modules/search-console/components/module/ModuleOverviewWidget/index.js:92]: I originally did but thought it might be cleaner to do this way since it needed the date range as well which it would need another prop for. I guess it really wasn't much of an improvement so I've moved it back into `Stats` with the added prop rather than changing the props it receives data from.

File: assets/js/util/partition-report.js (hunk 2 above)
Comment (aaemnnosttv) [reply assets/js/util/partition-report.js:45]: > Could this result in a potential problem that the compare date range would be incorrect (e.g. missing some days)?

I don't think so – it shouldn't be possible for the `compareRange` to be shorter than the `currentRange` since it is sliced from the beginning of the `report` array. Due to the way the two ranges are sliced before this, `currentRange` can never be longer because they both have the same length limit.

I suppose this makes the slicing on the `compareRange` here redundant and we could simply slice the `currentRange` to the length of the `compareRange` instead.

File: assets/js/util/partition-report.js (hunk 3 above)
Comment (aaemnnosttv) [reply assets/js/util/partition-report.js:34]: I've updated the range length argument to be `dateRangeLength` as you suggested – this cleaned up the consuming code as well 👍 

> The return object properties could possibly be something like `dateRangeEntries` and `compareDateRangeEntries`, just throwing some ideas out there.

This seems a bit more verbose than necessary. We should keep the return focused on the particular segment of the report.

We could also simply return an array from the function and leave the naming to the consumer via destructuring:

```js
const [ compare, current ] = partitionReport( data, { dateRangeLength } )
```

I wouldn't be opposed to this, but our use would be more consistent to return an object like it does now.

Comment (felixarntz) [reply assets/js/util/partition-report.js]: Using an object sounds good.

**Review by felixarntz — APPROVED** [review]
@aaemnnosttv I've added one more test here as mentioned that, while it doesn't really test anything extra, uses more realistic data to prove that the implementation maintains the weekday alignment when data is requested for a multitude of weeks. Let me know if that looks good to you / feel free to merge if so.

**Review by aaemnnosttv — COMMENTED** [review]
The added test LGTM – thanks for adding that @felixarntz 

I can't actually approve this since it's my own PR but ✅  from me 😄
